x/upgrade/internal/keeper/keeper.go

```diff
@@ -0,0 +1,119 @@
+package keeper
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/upgrade/exported"
+	"github.com/cosmos/cosmos-sdk/x/upgrade/internal/types"
+	"github.com/tendermint/tendermint/libs/log"
+)
+
+type Keeper struct {
+	storeKey        sdk.StoreKey
+	cdc             *codec.Codec
+	upgradeHandlers map[string]types.UpgradeHandler
+}
+
+var _ exported.Keeper = (*Keeper)(nil)
+
+// NewKeeper constructs an upgrade Keeper
+func NewKeeper(storeKey sdk.StoreKey, cdc *codec.Codec) *Keeper {
+	return &Keeper{
+		storeKey:        storeKey,
+		cdc:             cdc,
+		upgradeHandlers: map[string]types.UpgradeHandler{},
+	}
+}
+
+// SetUpgradeHandler sets an UpgradeHandler for the upgrade specified by name. This handler will be called when the upgrade
+// with this name is applied. In order for an upgrade with the given name to proceed, a handler for this upgrade
+// must be set even if it is a no-op function.
+func (k *Keeper) SetUpgradeHandler(name string, upgradeHandler types.UpgradeHandler) {
+	k.upgradeHandlers[name] = upgradeHandler
+}
+
+// ScheduleUpgrade schedules an upgrade based on the specified plan.
+// If there is another Plan already scheduled, it will overwrite it
+// (implicitly cancelling the current plan)
+func (k *Keeper) ScheduleUpgrade(ctx sdk.Context, plan types.Plan) sdk.Error {
+	err := plan.ValidateBasic()
+	if err != nil {
+		return err
+	}
+	if !plan.Time.IsZero() {
+		if !plan.Time.After(ctx.BlockHeader().Time) {
+			return sdk.ErrUnknownRequest("upgrade cannot be scheduled in the past")
+		}
+	} else if plan.Height <= ctx.BlockHeight() {
+		return sdk.ErrUnknownRequest("upgrade cannot be scheduled in the past")
+	}
+
+	store := ctx.KVStore(k.storeKey)
+	if store.Has(types.DoneHeightKey(plan.Name)) {
+		return sdk.ErrUnknownRequest(fmt.Sprintf("upgrade with name %s has already been completed", plan.Name))
+	}
+
+	bz := k.cdc.MustMarshalBinaryBare(plan)
+	store.Set(types.PlanKey(), bz)
+	return nil
+}
+
+func (k *Keeper) getDoneHeight(ctx sdk.Context, name string) int64 {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.DoneHeightKey(name))
```

Comment:
nit/non-blocking, but should start to use the `Prefix` store over adhoc key composition.

ref: https://github.com/cosmos/cosmos-sdk/issues/5254

Reply:
Updated